src/arch/cuda/hybridAPI/hapi_impl.cpp

```diff
@@ -199,6 +274,12 @@ void hapiMapping(char** argv) {
       device_count = CmiNodeSize(CmiMyNode());
     }
 
+    // Create a DeviceManager per GPU device
+    std::vector<DeviceManager>& device_managers = CsvAccess(gpu_manager).device_managers;
+    for (int i = 0; i < device_count; i++) {
+      device_managers.emplace_back(i, device_count * CmiMyNodeRankLocal() + i);
+    }
+
     // Count number of PEs per device
     CsvAccess(gpu_manager).pes_per_device = CmiNodeSize(CmiMyNode()) / device_count;
```

Comment:
Since this code is called by each PE, it looks like each PE of the node will be writing to this `csv` variable. Shouldn't it be written to by only PE 0?

Reply:
This block of code is only called by rank 0 PEs (so once per process).

Reply:
Ah okay, I see the `if (CmiMyRank() == 0) {` guard now. Thanks!